sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/StreamMetrics.scala

```diff
@@ -0,0 +1,244 @@
+    // as this may be different for each source if there are many sources in the query plan
+    // and getOffset is called serially on them.
+    if (previousInputIntervalOption.nonEmpty) {
+      sources.foreach { s =>
+        inputRates(s).update(numInputRows.getOrElse(s, 0), previousInputIntervalOption.get)
+      }
+    }
+
+    // Update processing rate = num rows processed for each source in current trigger duration
+    sources.foreach { s =>
+      processingRates(s).update(numInputRows.getOrElse(s, 0), currentTriggerDuration)
+    }
+
+    // Update output rate = num rows output to the sink in current trigger duration
+    outputRate.update(numOutputRows.getOrElse(0), currentTriggerDuration)
+    logDebug("Output rate updated to " + outputRate.currentRate)
+
+    // Update latency = if data present, 0.5 * previous trigger interval + current trigger duration
+    if (previousInputIntervalOption.nonEmpty && totalNumInputRows > 0) {
+      latency = Some((previousInputIntervalOption.get.toDouble / 2) + currentTriggerDuration)
+    } else {
+      latency = None
+    }
+
+    previousTriggerStartTimestamp = currentTriggerStartTimestamp
+    currentTriggerStartTimestamp = -1
+  }
+
+  // =========== Getter methods ===========
+
+  def currentInputRate(): Double = synchronized {
+    // Since we are calculating source input rates using the same time interval for all sources
+    // it is fine to calculate total input rate as the sum of per source input rate.
+    inputRates.map(_._2.currentRate).sum
+  }
+
+  def currentSourceInputRate(source: Source): Double = synchronized {
+    inputRates(source).currentRate
+  }
+
+  def currentProcessingRate(): Double = synchronized {
+    // Since we are calculating source processing rates using the same time interval for all sources
+    // it is fine to calculate total processing rate as the sum of per source processing rate.
+    processingRates.map(_._2.currentRate).sum
+  }
+
+  def currentSourceProcessingRate(source: Source): Double = synchronized {
+    processingRates(source).currentRate
+  }
+
+  def currentOutputRate(): Double = synchronized { outputRate.currentRate }
+
+  def currentLatency(): Option[Double] = synchronized { latency }
+
+  def currentTriggerStatus(): Map[String, String] = synchronized { triggerStatus.toMap }
+
+  def currentSourceTriggerStatus(source: Source): Map[String, String] = synchronized {
+    sourceTriggerStatus(source).toMap
+  }
+
+  // =========== Other methods ===========
+
+  private def registerGauge[T](name: String, f: () => T)(implicit num: Numeric[T]): Unit = {
+    synchronized {
+      metricRegistry.register(name, new Gauge[T] {
+        override def getValue: T = f()
+      })
+    }
+  }
+
+  def stop(): Unit = synchronized {
+    inputRates.valuesIterator.foreach { _.stop() }
+    processingRates.valuesIterator.foreach { _.stop() }
+    outputRate.stop()
+    latency = None
+  }
+}
+
+object StreamMetrics extends Logging {
+
+  class RateCalculator {
```

Comment:
I made that inner class because I wanted to wrap the simple state management of the rate calculation, without polluting the namespace with a class that is used only inside StreamMetrics.